src/array_api_extra/_funcs.py

```diff
@@ -548,3 +559,279 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    on read-only versus writeable arrays::
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
+
+    Warning
+    -------
+    The array API standard does not support integer array indices.
+    The behaviour of update methods when the index is an array of integers
+    is undefined; this is particularly true when the index contains multiple
+    occurrences of the same index, e.g. ``at(x, [0, 0]).set(2)``.
+
+    Note
+    ----
+    `sparse <https://sparse.pydata.org/>`_ is not supported by update methods yet.
+
+    See Also
+    --------
+    `jax.numpy.ndarray.at <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html>`_
+    """
+
+    x: Array
+    idx: Index
+    __slots__: ClassVar[tuple[str, str]] = ("idx", "x")
+
+    def __init__(self, x: Array, idx: Index = _undef, /):
+        self.x = x
+        self.idx = idx
+
+    def __getitem__(self, idx: Index) -> at:
+        """Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        if self.idx is not _undef:
+            msg = "Index has already been set"
+            raise ValueError(msg)
+        self.idx = idx
+        return self
+
+    def _common(
+        self,
+        at_op: str,
+        y: Array = _undef,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+        _is_update: bool = True,
+        **kwargs: Untyped,
+    ) -> tuple[Untyped, None] | tuple[None, Array]:
+        """Perform common prepocessing.
+
+        Returns
+        -------
+        If the operation can be resolved by at[], (return value, None)
+        Otherwise, (None, preprocessed x)
+        """
+        if self.idx is _undef:
+            msg = (
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+            raise TypeError(msg)
+
+        x = self.x
+
+        if copy is True:
+            writeable = None
+        elif copy is False:
+            writeable = is_writeable_array(x)
+            if not writeable:
+                msg = "Cannot modify parameter in place"
+                raise ValueError(msg)
+        elif copy is None:  # type: ignore[redundant-expr]
+            writeable = is_writeable_array(x)
+            copy = _is_update and not writeable
+        else:
+            msg = f"Invalid value for copy: {copy!r}"  # type: ignore[unreachable]  # pyright: ignore[reportUnreachable]
```

Comment:
is there a better way to deal with situations like this where invalid types can be passed at runtime?

Reply:
This double-ignore is a strong example of why I'm strongly in favour of having only one type checker. This is tedious to both read and write.

Reply:
it's definitely annoying, but unfortunately it's the only possibility for ensuring compatibility with these two main type-checkers

Reply:
but I agree with mypy and pyright that the `else` clause here is redundant

Reply:
you could alternatively move the `if copy is None` to the top, and then do `elif copy: ... else: ...`

Reply:
what if we still want to throw an error at runtime if `copy="foo"` is passed? I guess the argument is that, since this is internal, that would always be caught by the static analysis anyway?